src/theory/quantifiers/ceg_t_instantiator.cpp

```diff
@@ -1359,6 +1559,48 @@ Node BvInstantiator::rewriteTermForSolvePv(
                      children[1]));
     }
   }
+  else if (n.getKind() == EQUAL)
+  {
+    TNode lhs = n[0];
+    TNode rhs = n[1];
+
+    /* rewrite: x * x = x -> x < 2 */
+    if ((lhs == pv && rhs.getKind() == BITVECTOR_MULT && rhs[0] == pv
+         && rhs[1] == pv)
+        || (rhs == pv && lhs.getKind() == BITVECTOR_MULT && lhs[0] == pv
+            && lhs[1] == pv))
+    {
+      return nm->mkNode(
+          BITVECTOR_ULT,
+          pv,
+          bv::utils::mkConst(BitVector(bv::utils::getSize(pv), Integer(2))));
+    }
+
+    if (options::cbqiBvLinearize())
+    {
+      Node linear_n = linearizePvEq(pv, n, children, contains_pv);
+      if (!linear_n.isNull())
+        return linear_n;
+    }
+  }
+  else if (options::cbqiBvLinearize() && n.getKind() == BITVECTOR_PLUS
+           && contains_pv[n])
+  {
+    std::vector<Node> coeffs, leafs;
+    bool ok;
+
+    ok = collectPvCoeffs(pv, n, contains_pv, coeffs, leafs);
```

Comment:
We were going back and forth on this in the discussion, but after looking at this I do think we should simplify this.

My recommendation would be the following:

(1) Document the invariants on the what each of children[i] (the result of linearizing n[i]) can look like.  Adding the code gives something like:
- children[i] is of the form t1*pv + t2 if n[i] is an adder chain (with PLUS and NEG) that is linear for pv.

(2) Make collectPvCoeffs *not* traverse the structure of n, and instead be constructed by looking at children[i] directly, where the code is something like:
if children[i] = t1i*pv + t2i for all i=1...n, then we return (t11+...+t1n)*pv + (t21+...+t2n)
otherwise we return children[1]+....+children[n].

Notice the base case where children[i] = pv should maybe be a special case (otherwise we would have to rewrite pv to 1*pv+0 everywhere).

(3) Add a case below for (BV_NEG t) after this one, which would be something like:
If t is of the form t1*pv + t2, return (BV_NEG t1)*pv + (BV_NEG t2), otherwise return (BV_NEG t).

I believe this will make it much easier to add other linearization techniques if we follow this pattern.

Reply:
Further thoughts here, you could add:

(4) A case for BV_MUL, which would be:
if children[i] = t1i*pv + t2i for a single i \in { 1,...,n} and contains_pv[j] = false for j != i, then return (t1i*trem)*pv + (t2i*trem) where trem = children[1]*...*children[j-1]*children[j+1]*...*children[n]

Reply:
Btw, the functions you have I believe are pretty much what we need. So concretely, I would recommend simplifying collectPvCoeffs so that it does not traverse, and call it on children[1]....children[n] instead of n.